yjit/src/codegen.rs

```diff
@@ -1500,49 +1500,89 @@ fn gen_expandarray(
+        asm.stack_pop(1); // pop the array
         return Some(KeepCompiling);
     }
 
+    let array_opnd = asm.stack_opnd(0);
     let array_reg = asm.load(array_opnd);
     let array_len_opnd = get_array_len(asm, array_reg);
 
-    // Only handle the case where the number of values in the array is greater
-    // than or equal to the number of values requested.
-    asm.cmp(array_len_opnd, num.into());
-    asm.jl(Target::side_exit(Counter::expandarray_rhs_too_small));
+    // Guard on the comptime/expected array length
+    asm.comment(&format!("guard array length == {}", comptime_len));
+    asm.cmp(array_len_opnd, comptime_len.into());
+    jit_chain_guard(
+        JCC_JNE,
+        jit,
+        asm,
+        ocb,
+        OPT_AREF_MAX_CHAIN_DEPTH,
+        Counter::expandarray_chain_max_depth,
+    );
 
-    // Load the address of the embedded array into REG1.
-    // (struct RArray *)(obj)->as.ary
-    let array_reg = asm.load(array_opnd);
-    let ary_opnd = asm.lea(Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RARRAY_AS_ARY));
+    let array_opnd = asm.stack_pop(1); // pop after using the type info
 
-    // Conditionally load the address of the heap array into REG1.
-    // (struct RArray *)(obj)->as.heap.ptr
-    let flags_opnd = Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RBASIC_FLAGS);
-    asm.test(flags_opnd, Opnd::UImm(RARRAY_EMBED_FLAG as u64));
-    let heap_ptr_opnd = Opnd::mem(
-        usize::BITS as u8,
-        asm.load(array_opnd),
-        RUBY_OFFSET_RARRAY_AS_HEAP_PTR,
-    );
-    let ary_opnd = asm.csel_nz(ary_opnd, heap_ptr_opnd);
+    //println!("generating loads");
 
-    // Loop backward through the array and push each element onto the stack.
-    for i in (0..num).rev() {
-        let top = asm.stack_push(Type::Unknown);
-        let offset = i32::try_from(i * SIZEOF_VALUE).unwrap();
-        asm.mov(top, Opnd::mem(64, ary_opnd, offset));
+    // If the array has length 0, then we don't even need the array pointer
+    if comptime_len == 0 {
+        // Loop backward through the array and push Qnils onto the stack.
+        for _ in 0..num {
+            let top = asm.stack_push(Type::Nil);
+            asm.mov(top, Qnil.into());
+        }
+    } else {
+        let array_reg = asm.load(array_opnd);
+        let ary_opnd = asm.lea(Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RARRAY_AS_ARY));
+
+        // Conditionally load the address of the heap array
+        // (struct RArray *)(obj)->as.heap.ptr
+        let flags_opnd = Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RBASIC_FLAGS);
+        asm.test(flags_opnd, Opnd::UImm(RARRAY_EMBED_FLAG as u64));
+        let heap_ptr_opnd = Opnd::mem(
+            usize::BITS as u8,
+            asm.load(array_opnd),
+            RUBY_OFFSET_RARRAY_AS_HEAP_PTR,
+        );
+        let ary_opnd = asm.csel_nz(ary_opnd, heap_ptr_opnd);
+
+        // Loop backward through the array and push each element onto the stack.
+        for i in (0..num).rev() {
```

Comment:
I wonder if it makes sense to share this loop for `comptime_len == 0` case as well, e.g.

```rs
let ary_opnd = if comptime_len > 0 {
    // ...
    Some(asm.csel_nz(ary_opnd, heap_ptr_opnd))
} else {
    None
};

for i in (0..num).rev() {
    // ...
    let elem_opnd = if (i as u32) < comptime_len { Opnd::mem(64, ary_opnd.unwrap(), offset) } else { Qnil.into() };
    asm.mov(top, elem_opnd);
}
```